### PR TITLE
Use strongly_connected_components for Matrix

### LINKED ISSUE #21470: Use strongly_connected_components for block triangulation of matrix
I want to complete the work https://github.com/sympy/sympy/pull/19045, https://github.com/sympy/sympy/issues/16207

It is possible to use `strongly_connected_components` to block triangulate the matrix in a completely identical way to block diagonalizing the matrix
Although block triangulation looks less obvious to use than block diagonalization, 
it has some usage because solving block triangulated linear system is more efficient, 
and also it can give more granular decomposition than block diagonalization for nonsymmetric matrices.

This can also be made for some internal `DomainMatrix`, however, since we don't have detailed plan for block matrix features for `DomainMatrix`, it would not look very obvious to implement for now.

```python3
from sympy import *
from sympy.combinatorics.permutations import Permutation
from sympy.matrices.expressions.blockmatrix import BlockMatrix
from sympy.matrices.expressions.permutation import PermutationMatrix


def strongly_connected_components(M):
    from sympy.utilities.iterables import strongly_connected_components
    
    if not M.is_square:
        raise NonSquareMatrixError

    V = range(M.rows)
    E = sorted(M.todok().keys())
    return strongly_connected_components((V, E))


def strongly_connected_components_decomposition(M):
    iblocks = strongly_connected_components(M)

    p = Permutation(flatten(iblocks))
    P = PermutationMatrix(p)

    rows = []
    for a in iblocks:
        cols = []
        for b in iblocks:
            cols.append(M[a, b])
        rows.append(cols)
    B = BlockMatrix(rows)
    return P, B
```

Usage:
```python3
>>> A = MatrixSymbol('A', 9, 9).as_explicit()
>>> mask = BlockMatrix(
...     [
...         [Matrix.ones(3, 3), Matrix.zeros(3, 3), Matrix.ones(3, 3)],
...         [Matrix.zeros(3, 3), Matrix.ones(3, 3), Matrix.zeros(3, 3)],
...         [Matrix.zeros(3, 3), Matrix.zeros(3, 3), Matrix.ones(3, 3)]
...     ]
... )
>>> mask = mask.as_explicit()
>>> P = PermutationMatrix(Permutation.random(9)).as_explicit()
>>> A = A.multiply_elementwise(mask)
>>> pprint(A, use_unicode=False)
[A_00  A_01  A_02   0     0     0    A_06  A_07  A_08]
[                                                    ]
[A_10  A_11  A_12   0     0     0    A_16  A_17  A_18]
[                                                    ]
[A_20  A_21  A_22   0     0     0    A_26  A_27  A_28]
[                                                    ]
[ 0     0     0    A_33  A_34  A_35   0     0     0  ]
[                                                    ]
[ 0     0     0    A_43  A_44  A_45   0     0     0  ]
[                                                    ]
[ 0     0     0    A_53  A_54  A_55   0     0     0  ]
[                                                    ]
[ 0     0     0     0     0     0    A_66  A_67  A_68]
[                                                    ]
[ 0     0     0     0     0     0    A_76  A_77  A_78]
[                                                    ]
[ 0     0     0     0     0     0    A_86  A_87  A_88]
>>> B = P*A*P.T
>>> pprint(B, use_unicode=False)
[A_33   0     0    A_35  A_34   0     0     0     0  ]
[                                                    ]
[ 0    A_11  A_12   0     0    A_17  A_18  A_16  A_10]
[                                                    ]
[ 0    A_21  A_22   0     0    A_27  A_28  A_26  A_20]
[                                                    ]
[A_53   0     0    A_55  A_54   0     0     0     0  ]
[                                                    ]
[A_43   0     0    A_45  A_44   0     0     0     0  ]
[                                                    ]
[ 0     0     0     0     0    A_77  A_78  A_76   0  ]
[                                                    ]
[ 0     0     0     0     0    A_87  A_88  A_86   0  ]
[                                                    ]
[ 0     0     0     0     0    A_67  A_68  A_66   0  ]
[                                                    ]
[ 0    A_01  A_02   0     0    A_07  A_08  A_06  A_00]
>>> print(strongly_connected_components(B))
[[0, 3, 4], [5, 6, 7], [1, 2, 8]]
>>> P, A = strongly_connected_components_decomposition(B)
>>> print(P)
PermutationMatrix((8)(1 3 5 7 2 4 6))
>>> pprint(A, use_unicode=False)
[[A_33  A_35  A_34]      [0  0  0]           [0  0  0]     ]
[[                ]      [       ]           [       ]     ]
[[A_53  A_55  A_54]      [0  0  0]           [0  0  0]     ]
[[                ]      [       ]           [       ]     ]
[[A_43  A_45  A_44]      [0  0  0]           [0  0  0]     ]
[                                                          ]
[    [0  0  0]       [A_77  A_78  A_76]      [0  0  0]     ]
[    [       ]       [                ]      [       ]     ]
[    [0  0  0]       [A_87  A_88  A_86]      [0  0  0]     ]
[    [       ]       [                ]      [       ]     ]
[    [0  0  0]       [A_67  A_68  A_66]      [0  0  0]     ]
[                                                          ]
[    [0  0  0]       [A_17  A_18  A_16]  [A_11  A_12  A_10]]
[    [       ]       [                ]  [                ]]
[    [0  0  0]       [A_27  A_28  A_26]  [A_21  A_22  A_20]]
[    [       ]       [                ]  [                ]]
[    [0  0  0]       [A_07  A_08  A_06]  [A_01  A_02  A_00]]
```

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #21470 

#### Brief description of what is fixed or changed

In this pass, I'm attempting to use `strongly_connected_components` for places like `eigenvals`, `det` to replace the triangular heuristics and to replace the already present improvement done by `connected_components`.

~For computing `charpoly`, I took the reverse approach by computing the multiplication of polynomial
However, I don't think that any method using determinant or eigenvalues relies directly on `charpoly` after this, so it would just be an individual improvement. (Since there is no class in sympy that can correspond to `PolyMul`, I don't think that it can use the benefit of already computed factorization)
I may rather disentangle the `charpoly` to avoid doubly computing `strongly_connected_components`~
I won't attempt to improve `charpoly` here since it looks like messing up the organization of the module

#### Other comments

I would also want to correct the documentation `A = P B P^{-1}` to `A = P^{-1} B P` since I discovered that was mathematically incorrect by testing with random matrix.

I would also investigate if giving upper triangular decomposition is easy to build on top of this framework (like with transposition) or has some mathematical difficulty such that we have reason to provide upper or lower option.
The only reason that this gives lower triangular result is because it corresponds to the output of `iterable.strongly_connected_components`, but since lots of matrix methods like `rref` are based on upper triangulation, we need a way to convert it to upper triangular decomposition.

#### Release Notes

- matrices
  - Added `Matrix.strongly_connected_components` and `Matrix.strongly_connected_components_decomposition` for computing block triangulation of the matrix.

### CODE REVIEW

File: sympy/matrices/graph.py [hunk 1]
```diff
@@ -37,6 +36,15 @@ def _connected_components(M):
     return connected_components((V, E))
 
 
+def _strongly_connected_components(M):
+    if not M.is_square:
+        raise NonSquareMatrixError
+
+    V = range(M.rows)
+    E = sorted(M.todok().keys())
+    return strongly_connected_components((V, E))
```
Comment (oscarbenjamin) [inline sympy/matrices/graph.py:45]: The `todok` function should be made faster:
```python
In [3]: for n in [1, 10, 100, 1000]:
   ...:     %time eye(n).det()
   ...: 
CPU times: user 106 µs, sys: 1 µs, total: 107 µs
Wall time: 115 µs
CPU times: user 3.23 ms, sys: 434 µs, total: 3.66 ms
Wall time: 4.48 ms
CPU times: user 151 ms, sys: 20.4 ms, total: 171 ms
Wall time: 202 ms
CPU times: user 9.67 s, sys: 77.4 ms, total: 9.74 s
Wall time: 9.86 s

In [6]: %prun -s cumulative eye(1000).det()

         20102028 function calls in 19.422 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   19.422   19.422 {built-in method builtins.exec}
        1    0.003    0.003   19.422   19.422 <string>:1(<module>)
        1    0.000    0.000   19.412   19.412 matrices.py:127(det)
        1    0.002    0.002   19.412   19.412 determinant.py:548(_det)
        1    0.000    0.000   19.338   19.338 matrices.py:2210(strongly_connected_components)
        1    0.000    0.000   19.338   19.338 graph.py:39(_strongly_connected_components)
        1    0.000    0.000   19.314   19.314 common.py:619(todok)
        1    1.572    1.572   19.314   19.314 common.py:180(_eval_todok)
  1002000    1.829    0.000   15.382    0.000 dense.py:62(__getitem__)
  1002000    4.964    0.000   13.300    0.000 matrices.py:1819(key2ij)
  1002000    1.976    0.000    3.981    0.000 matrices.py:1834(<listcomp>)
  1006000    1.064    0.000    3.647    0.000 compatibility.py:208(is_sequence)
  1000000    0.752    0.000    2.424    0.000 numbers.py:2242(__ne__)
  1006000    1.191    0.000    2.209    0.000 compatibility.py:150(iterable)
  6019000    1.968    0.000    1.968    0.000 {built-in method builtins.isinstance}
  1000000    0.908    0.000    1.672    0.000 numbers.py:2235(__eq__)
  2005000    1.492    0.000    1.492    0.000 common.py:3239(a2idx)
  2012000    0.633    0.000    0.633    0.000 {built-in method builtins.hasattr}
  1002001    0.484    0.000    0.484    0.000 common.py:601(shape)
  1007000    0.287    0.000    0.287    0.000 {built-in method builtins.iter}
  1006001    0.235    0.000    0.235    0.000 {built-in method builtins.len}
     1000    0.015    0.000    0.033    0.000 common.py:317(extract)
        1    0.003    0.003    0.024    0.024 iterables.py:1046(strongly_connected_components)
```

File: sympy/matrices/determinant.py [hunk 2]
```diff
@@ -669,14 +661,17 @@ def _det(M, method="bareiss", iszerofunc=None):
...
+    for b in M.strongly_connected_components():
+        if method == "domain-ge": # uses DomainMatrix to evalute determinant
+            det *= _det_DOM(M[b, b])
+        elif method == "bareiss":
+            det *= M[b, b]._eval_det_bareiss(iszerofunc=iszerofunc)
+        elif method == "berkowitz":
+            det *= M[b, b]._eval_det_berkowitz()
+        elif method == "lu":
+            det *= M[b, b]._eval_det_lu(iszerofunc=iszerofunc)
+    return det
```
Comment (oscarbenjamin) [inline sympy/matrices/determinant.py:674]: Currently `todok` is too slow to notice this but creating a product with repeated use of `*` is very slow compared to collecting all factors and calling `Mul` once. The `Mul.flatten()` routine is `O(N)` so repeated use of `*` is `O(N**2)`:
```python
In [19]: def prod1(n):
    ...:     d = S.One
    ...:     for _ in range(n):
    ...:         d *= Dummy()
    ...:     return d
    ...: 

In [20]: def prod2(n):
    ...:     factors = [Dummy() for _ in range(n)]
    ...:     return Mul(*factors)
    ...: 

In [21]: prod1(3)
Out[21]: Dummy₂₆⋅Dummy₂₇⋅Dummy₂₈

In [22]: prod2(3)
Out[22]: Dummy₂₉⋅Dummy₃₀⋅Dummy₃₁

In [23]: for n in [1, 10, 100]:
    ...:     %time prod1(n)
    ...: 
CPU times: user 142 µs, sys: 9 µs, total: 151 µs
Wall time: 157 µs
CPU times: user 8.05 ms, sys: 87 µs, total: 8.14 ms
Wall time: 8.31 ms
CPU times: user 347 ms, sys: 1.9 ms, total: 349 ms
Wall time: 351 ms

In [24]: for n in [1, 10, 100]:
    ...:     %time prod2(n)
    ...: 
CPU times: user 166 µs, sys: 11 µs, total: 177 µs
Wall time: 194 µs
CPU times: user 860 µs, sys: 3 µs, total: 863 µs
Wall time: 874 µs
CPU times: user 5.21 ms, sys: 88 µs, total: 5.3 ms
Wall time: 5.52 ms
```

File: sympy/matrices/graph.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/matrices/graph.py:45]:
```diff
diff --git a/sympy/matrices/common.py b/sympy/matrices/common.py
index 5e82997798..32bd2691ed 100644
--- a/sympy/matrices/common.py
+++ b/sympy/matrices/common.py
@@ -178,14 +178,8 @@ def _eval_tolist(self):
         return [list(self[i,:]) for i in range(self.rows)]
 
     def _eval_todok(self):
-        dok = {}
-        rows, cols = self.shape
-        for i in range(rows):
-            for j in range(cols):
-                val = self[i, j]
-                if val != self.zero:
-                    dok[i, j] = val
-        return dok
+        cols = self.cols
+        return {divmod(ij, cols): e for ij, e in enumerate(self._mat) if e}
 
     def _eval_vec(self):
         rows = self.rows
```

**Comment by sylee957** [comment]
We need to have some additional benchmark examples about these
The first one to test out is how additionally this improves cases for random block diagonal matrices
And the second one is to test out how additionally this doesn't slow down than triangular heuristics that was present before.

**Comment by oscarbenjamin** [comment]
> We need to have some additional benchmark examples about these

It would be good to have tests for cases like this in the benchmark suite. I don't think that we need to go too far with this though because if Matrix was made to use DomainMatrix which uses sparse SDM internally then computing the strongly connected components could be much faster.

**Comment by oscarbenjamin** [comment]
Looks good to me

**Comment by sylee957** [comment]
I will write some documentation before merge

**Pull request merged**